p42/src/bank.rs

```diff
@@ -0,0 +1,154 @@
+    pub balance: i64,
+}
+
+#[derive(Debug, Clone)]
+/// A struct representing a bank with a list of users, a name, and interest rates.
+pub struct Bank {
+    pub users: HashMap<String, User>,
+    pub name: String,
+    pub credit_interest: u64,
+    pub debit_interest: u64,
+}
+
+impl Bank {
+    /// Creates a new bank with the given name, credit interest, and debit interest.
+    pub fn new(name: String, credit_interest: u64, debit_interest: u64) -> Self {
+        Bank {
+            users: HashMap::new(),
+            name,
+            credit_interest,
+            debit_interest,
+        }
+    }
+
+    /// Adds a user to the bank with the given name and credit line.
+    pub fn add_user(&mut self, name: String, credit_line: u64) {
+        let user = User {
+            name,
+            credit_line,
+            balance: 0,
+        };
+        self.users.insert(user.name.clone(), user);
+    }
+
+    /// Gets a user by their name.
+    pub fn get_user(&self, name: &str) -> Option<&User> {
+        self.users.get(name)
+    }
+
+    /// Gets a mutable reference to a user by their name.
+    pub fn get_user_mut(&mut self, name: &str) -> Option<&mut User> {
+        self.users.get_mut(name)
+    }
+
+    /// Gets the total assets and total liabilities for the bank.
+    pub fn calc_balance(&self) -> (i64, i64) {
+        let mut total_liabilities = 0i64;
+        let mut total_assets = 0i64;
+
+        for user in self.users.values() {
+            if user.balance < 0 {
+                total_assets = total_assets
+                    .checked_add(user.balance)
+                    .ok_or("Overflow in total assets")
+                    .unwrap();
+            } else {
+                total_liabilities = total_liabilities
+                    .checked_add(user.balance)
+                    .ok_or("Overflow in total liabilities")
+                    .unwrap();
+            }
+        }
+
+        (total_liabilities, total_assets)
+    }
+
+    /// Transfers amount from one user to another.
+    pub fn transfer_funds(
+        &mut self,
+        from_user: &str,
+        to_user: &str,
+        amount: u64,
+    ) -> Result<(), String> {
+        let from = self.get_user(from_user).ok_or("From user not found")?;
+
+        if from.balance + i64::try_from(from.credit_line).expect("Unable to convert balance to i64")
+            < i64::try_from(amount).expect("Unable to convert amount to i64")
+        {
+            return Err("Insufficient credit limit".to_string());
+        }
+        let from_balance = from
+            .balance
+            .checked_sub(i64::try_from(amount).expect("Unable to convert amount to i64"))
+            .ok_or("Underflow in balance")
+            .unwrap();
+
+        let to_mut = self.get_user_mut(to_user).ok_or("To user not found")?;
+        let to_balance = to_mut
+            .balance
+            .checked_add(i64::try_from(amount).expect("Unable to convert amount to i64"))
```

Comment:
You convert `amount` several times to `i64`. Instead it's better to do the conversion once and use the resulting variable.